File: module_16 - The Complete Compiler/parser.c
/*
 * parser.c — Recursive-descent parser for mycc
 * Module 16: The Complete Compiler
 *
 * Grammar (simplified):
 *   program    := func*
 *   func       := type IDENT '(' params ')' block
 *   params     := ε | type IDENT (',' type IDENT)*
 *   block      := '{' stmt* '}'
 *   stmt       := var_decl | return_stmt | if_stmt | while_stmt
 *               | print_stmt | expr_stmt
 *   expr_stmt  := expr ';'
 *   expr       := assign
 *   assign     := IDENT '=' assign | or_expr
 *   or_expr    := and_expr ('||' and_expr)*
 *   and_expr   := eq_expr  ('&&' eq_expr)*
 *   eq_expr    := rel_expr (('=='|'!=') rel_expr)*
 *   rel_expr   := add_expr (('<'|'<='|'>'|'>=') add_expr)*
 *   add_expr   := mul_expr (('+' | '-') mul_expr)*
 *   mul_expr   := unary   (('*'|'/'|'%') unary)*
 *   unary      := '-' unary | '!' unary | primary
 *   primary    := INT_LIT | IDENT | IDENT '(' args ')' | '(' expr ')'
 */
#include "parser.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* ------------------------------------------------------------------ helpers */

static void error(Parser *p, const char *msg)
{
    fprintf(stderr, "%d:%d: parse error: %s (got '%.*s')\n",
            p->cur.line, p->cur.col, msg,
            (int)p->cur.len, p->cur.start);
    p->errors++;
}

static Token advance(Parser *p)
{
    Token prev = p->cur;
    p->cur = lexer_next(p->lex);
    return prev;
}

static int check(Parser *p, TokenType t)
{
    return p->cur.type == t;
}

static int match(Parser *p, TokenType t)
{
    if (!check(p, t)) return 0;
    advance(p);
    return 1;
}

static Token expect(Parser *p, TokenType t)
{
    if (!check(p, t)) {
        char buf[64];
        snprintf(buf, sizeof buf, "expected '%s'", token_type_name(t));
        error(p, buf);
    }
    return advance(p);
}

/* ----------------------------------------------------------------- forward decls */
static ASTNode *parse_stmt(Parser *p);
static ASTNode *parse_expr(Parser *p);
static ASTNode *parse_block(Parser *p);

/* ----------------------------------------------------------------- expressions */

static char *dup_tok(const Token *t)
{
    char *s = (char *)malloc(t->len + 1);
    if (!s) { fprintf(stderr, "parser: oom\n"); exit(1); }
    memcpy(s, t->start, t->len);
    s[t->len] = '\0';
    return s;
}

static ASTNode *parse_primary(Parser *p)
{
    int line = p->cur.line;

    if (check(p, TOK_INT_LIT)) {
        Token t = advance(p);
        ASTNode *n = ast_new(AST_INT_LIT, line);
        n->u.ival = t.ival;
        return n;
    }

    if (check(p, TOK_IDENT)) {
        Token t = advance(p);
        /* function call? */
        if (check(p, TOK_LPAREN)) {
            advance(p); /* eat ( */
            ASTNode *n = ast_new(AST_CALL, line);
            n->u.call.name = dup_tok(&t);
            n->u.call.args = NULL;
            n->u.call.argc = 0;
            int cap = 0;
            if (!check(p, TOK_RPAREN)) {
                do {
                    if (n->u.call.argc == cap) {
                        cap = cap ? cap * 2 : 4;
                        n->u.call.args = realloc(n->u.call.args, (size_t)cap * sizeof(ASTNode*));
                    }
                    n->u.call.args[n->u.call.argc++] = parse_expr(p);
                } while (match(p, TOK_COMMA));
            }
            expect(p, TOK_RPAREN);
            return n;
        }
        /* plain identifier */
        ASTNode *n = ast_new(AST_IDENT, line);
        n->u.name = dup_tok(&t);
        return n;
    }

    if (match(p, TOK_LPAREN)) {
        ASTNode *inner = parse_expr(p);
        expect(p, TOK_RPAREN);
        return inner;
    }

    error(p, "expected expression");
    /* return a dummy node so parsing can continue */
    ASTNode *dummy = ast_new(AST_INT_LIT, line);
    dummy->u.ival = 0;
    return dummy;
}

static ASTNode *parse_unary(Parser *p)
{
    int line = p->cur.line;
    if (check(p, TOK_MINUS) || check(p, TOK_NOT)) {
        Token op = advance(p);
        ASTNode *n = ast_new(AST_UNOP, line);
        n->u.binop.op    = (int)op.type;
        n->u.binop.left  = parse_unary(p);
        n->u.binop.right = NULL;
        return n;
    }
    return parse_primary(p);
}

static ASTNode *parse_mul(Parser *p)
{
    ASTNode *left = parse_unary(p);
    while (check(p, TOK_STAR) || check(p, TOK_SLASH) || check(p, TOK_PERCENT)) {
        Token op = advance(p);
        ASTNode *n = ast_new(AST_BINOP, left->line);
        n->u.binop.op    = (int)op.type;
        n->u.binop.left  = left;
        n->u.binop.right = parse_unary(p);
        left = n;
    }
    return left;
}

static ASTNode *parse_add(Parser *p)
{
    ASTNode *left = parse_mul(p);
    while (check(p, TOK_PLUS) || check(p, TOK_MINUS)) {
        Token op = advance(p);
        ASTNode *n = ast_new(AST_BINOP, left->line);
        n->u.binop.op    = (int)op.type;
        n->u.binop.left  = left;
        n->u.binop.right = parse_mul(p);
        left = n;
    }
    return left;
}

static ASTNode *parse_rel(Parser *p)
{
    ASTNode *left = parse_add(p);
    while (check(p, TOK_LT) || check(p, TOK_LE) ||
           check(p, TOK_GT) || check(p, TOK_GE)) {
        Token op = advance(p);
        ASTNode *n = ast_new(AST_BINOP, left->line);
        n->u.binop.op    = (int)op.type;
        n->u.binop.left  = left;
        n->u.binop.right = parse_add(p);
        left = n;
    }
    return left;
}

static ASTNode *parse_eq(Parser *p)
{
    ASTNode *left = parse_rel(p);
    while (check(p, TOK_EQ) || check(p, TOK_NEQ)) {
        Token op = advance(p);
        ASTNode *n = ast_new(AST_BINOP, left->line);
        n->u.binop.op    = (int)op.type;
        n->u.binop.left  = left;
        n->u.binop.right = parse_rel(p);
        left = n;
    }
    return left;
}

static ASTNode *parse_and(Parser *p)
{
    ASTNode *left = parse_eq(p);
    while (check(p, TOK_AND)) {
        Token op = advance(p);
        ASTNode *n = ast_new(AST_BINOP, left->line);
        n->u.binop.op    = (int)op.type;
        n->u.binop.left  = left;
        n->u.binop.right = parse_eq(p);
        left = n;
    }
    return left;
}

static ASTNode *parse_or(Parser *p)
{
    ASTNode *left = parse_and(p);
    while (check(p, TOK_OR)) {
        Token op = advance(p);
        ASTNode *n = ast_new(AST_BINOP, left->line);
        n->u.binop.op    = (int)op.type;
        n->u.binop.left  = left;
        n->u.binop.right = parse_and(p);
        left = n;
    }
    return left;
}

static ASTNode *parse_assign(Parser *p)
{
    /* Peek: if IDENT followed by '=', it's assignment */
    if (check(p, TOK_IDENT)) {
        Lexer saved = *p->lex;
        Token cur_saved = p->cur;

        Token ident = advance(p);
        if (check(p, TOK_ASSIGN)) {
            advance(p); /* eat = */
            ASTNode *n = ast_new(AST_ASSIGN, ident.line);
            n->u.assign.name  = dup_tok(&ident);
            n->u.assign.value = parse_assign(p);
            return n;
        }
        /* Not assignment — restore and parse as or_expr */
        *p->lex = saved;
        p->cur  = cur_saved;
    }
    return parse_or(p);
}

static ASTNode *parse_expr(Parser *p)
{
    return parse_assign(p);
}

/* ----------------------------------------------------------------- statements */

static ASTNode *parse_stmt(Parser *p)
{
    int line = p->cur.line;

    /* var declaration: int name; */
    if (check(p, TOK_KW_INT)) {
        advance(p);
        Token name = expect(p, TOK_IDENT);
        expect(p, TOK_SEMICOLON);
        ASTNode *n = ast_new(AST_VAR_DECL, line);
        n->u.name = dup_tok(&name);
        return n;
    }

    /* return */
    if (match(p, TOK_KW_RETURN)) {
        ASTNode *n = ast_new(AST_RETURN, line);
        if (check(p, TOK_SEMICOLON)) {
            advance(p);
            n->u.expr = NULL;
        } else {
            n->u.expr = parse_expr(p);
            expect(p, TOK_SEMICOLON);
        }
        return n;
    }

    /* if */
    if (match(p, TOK_KW_IF)) {
        expect(p, TOK_LPAREN);
        ASTNode *cond = parse_expr(p);
        expect(p, TOK_RPAREN);
        ASTNode *then_b = parse_stmt(p);
        ASTNode *else_b = NULL;
        if (match(p, TOK_KW_ELSE))
            else_b = parse_stmt(p);
        ASTNode *n = ast_new(AST_IF, line);
        n->u.if_stmt.cond        = cond;
        n->u.if_stmt.then_branch = then_b;
        n->u.if_stmt.else_branch = else_b;
        return n;
    }

    /* while */
    if (match(p, TOK_KW_WHILE)) {
        expect(p, TOK_LPAREN);
        ASTNode *cond = parse_expr(p);
        expect(p, TOK_RPAREN);
        ASTNode *body = parse_stmt(p);
        ASTNode *n = ast_new(AST_WHILE, line);
        n->u.while_stmt.cond = cond;
        n->u.while_stmt.body = body;
        return n;
    }

    /* print(expr); */
    if (match(p, TOK_KW_PRINT)) {
        expect(p, TOK_LPAREN);
        ASTNode *val = parse_expr(p);
        expect(p, TOK_RPAREN);
        expect(p, TOK_SEMICOLON);
        ASTNode *n = ast_new(AST_PRINT, line);
        n->u.expr = val;
        return n;
    }

    /* block */
    if (check(p, TOK_LBRACE))
        return parse_block(p);

    /* expression statement */
    ASTNode *expr = parse_expr(p);
    expect(p, TOK_SEMICOLON);
    ASTNode *n = ast_new(AST_EXPR_STMT, line);
    n->u.expr = expr;
    return n;
}

static ASTNode *parse_block(Parser *p)
{
    int line = p->cur.line;
    expect(p, TOK_LBRACE);
    ASTNode *n = ast_new(AST_BLOCK, line);
    n->u.block.stmts = NULL;
    n->u.block.count = 0;
    int cap = 0;
    while (!check(p, TOK_RBRACE) && !check(p, TOK_EOF)) {
        if (n->u.block.count == cap) {
            cap = cap ? cap * 2 : 8;
            n->u.block.stmts = realloc(n->u.block.stmts,
                                       (size_t)cap * sizeof(ASTNode*));
        }
        n->u.block.stmts[n->u.block.count++] = parse_stmt(p);
    }
    expect(p, TOK_RBRACE);
    return n;
}

/* ----------------------------------------------------------------- functions / program */

static int is_type(Parser *p)
{
    return check(p, TOK_KW_INT) || check(p, TOK_KW_VOID);
}

static ASTNode *parse_func(Parser *p)
{
    int line = p->cur.line;
    /* return type */
    if (!is_type(p)) { error(p, "expected type"); return NULL; }
    advance(p);

    Token name = expect(p, TOK_IDENT);
    expect(p, TOK_LPAREN);

    char **params = NULL;
    int param_count = 0;
    int param_cap   = 0;

    if (!check(p, TOK_RPAREN)) {
        do {
            if (!is_type(p)) { error(p, "expected parameter type"); break; }
            advance(p); /* eat type */
            Token pname = expect(p, TOK_IDENT);
            if (param_count == param_cap) {
                param_cap = param_cap ? param_cap * 2 : 4;
                params = realloc(params, (size_t)param_cap * sizeof(char*));
            }
            params[param_count++] = dup_tok(&pname);
        } while (match(p, TOK_COMMA));
    }
    expect(p, TOK_RPAREN);

    ASTNode *body = parse_block(p);

    ASTNode *n = ast_new(AST_FUNC, line);
    n->u.func.name        = dup_tok(&name);
    n->u.func.params      = params;
    n->u.func.param_count = param_count;
    n->u.func.body        = body;
    return n;
}

void parser_init(Parser *p, Lexer *lex)
{
    p->lex    = lex;
    p->errors = 0;
    p->cur    = lexer_next(lex);
}

ASTNode *parse_program(Parser *p)
{
    ASTNode *prog = ast_new(AST_PROGRAM, 1);
    prog->u.program.funcs = NULL;
    prog->u.program.count = 0;
    int cap = 0;

    while (!check(p, TOK_EOF)) {
        ASTNode *fn = parse_func(p);
        if (!fn) break;
        if (prog->u.program.count == cap) {
            cap = cap ? cap * 2 : 4;
            prog->u.program.funcs = realloc(prog->u.program.funcs,
                                            (size_t)cap * sizeof(ASTNode*));
        }
        prog->u.program.funcs[prog->u.program.count++] = fn;
    }
    return prog;
}

File: webapp/src/parser.ts
/*
 * parser.ts — TypeScript port of a-finity's real recursive-descent parser.
 *
 * Ported directly from module_16 - The Complete Compiler/parser.c.
 * Grammar (verbatim from the header comment in parser.c):
 *
 *   program    := func*
 *   func       := type IDENT '(' params ')' block
 *   params     := ε | type IDENT (',' type IDENT)*
 *   block      := '{' stmt* '}'
 *   stmt       := var_decl | return_stmt | if_stmt | while_stmt
 *               | print_stmt | expr_stmt
 *   expr_stmt  := expr ';'
 *   expr       := assign
 *   assign     := IDENT '=' assign | or_expr
 *   or_expr    := and_expr ('||' and_expr)*
 *   and_expr   := eq_expr  ('&&' eq_expr)*
 *   eq_expr    := rel_expr (('=='|'!=') rel_expr)*
 *   rel_expr   := add_expr (('<'|'<='|'>'|'>=') add_expr)*
 *   add_expr   := mul_expr (('+' | '-') mul_expr)*
 *   mul_expr   := unary   (('*'|'/'|'%') unary)*
 *   unary      := '-' unary | '!' unary | primary
 *   primary    := INT_LIT | IDENT | IDENT '(' args ')' | '(' expr ')'
 *
 * Every parse_* function below mirrors its C counterpart function-for-
 * function, including the two-token-lookahead trick parse_assign() uses
 * (save lexer + current token, try IDENT '=', restore if it doesn't match).
 */

import { Lexer, Token, TokenType, TOKEN_TYPE_NAME } from './lexer';
import type { ASTNode } from './ast';

export interface ParseError {
  message: string;
  line: number;
  col: number;
}

export class Parser {
  private lex: Lexer;
  private cur: Token;
  readonly errors: ParseError[] = [];

  constructor(lex: Lexer) {
    this.lex = lex;
    this.cur = this.lex.next();
  }

  private error(msg: string): void {
    this.errors.push({ message: msg, line: this.cur.line, col: this.cur.col });
  }

  private advance(): Token {
    const prev = this.cur;
    this.cur = this.lex.next();
    return prev;
  }

  private check(t: TokenType): boolean {
    return this.cur.type === t;
  }

  private match(t: TokenType): boolean {
    if (!this.check(t)) return false;
    this.advance();
    return true;
  }

  private expect(t: TokenType): Token {
    if (!this.check(t)) {
      this.error(`expected '${TOKEN_TYPE_NAME[t]}' but got '${this.cur.lexeme || TOKEN_TYPE_NAME[this.cur.type]}'`);
    }
    return this.advance();
  }

  // ----------------------------------------------------------------- expressions

  private parsePrimary(): ASTNode {
    const line = this.cur.line;

    if (this.check(TokenType.INT_LIT)) {
      const t = this.advance();
      return { kind: 'INT_LIT', line, ival: t.ival };
    }

    if (this.check(TokenType.IDENT)) {
      const t = this.advance();
      // function call?
      if (this.check(TokenType.LPAREN)) {
        this.advance(); // eat (
        const args: ASTNode[] = [];
        if (!this.check(TokenType.RPAREN)) {
          do {
            args.push(this.parseExpr());
          } while (this.match(TokenType.COMMA));
        }
        this.expect(TokenType.RPAREN);
        return { kind: 'CALL', line, name: t.lexeme, args };
      }
      // plain identifier
      return { kind: 'IDENT', line, name: t.lexeme };
    }

    if (this.match(TokenType.LPAREN)) {
      const inner = this.parseExpr();
      this.expect(TokenType.RPAREN);
      return inner;
    }

    this.error('expected expression');
    // dummy node so parsing can continue, matching parser.c's recovery
    return { kind: 'INT_LIT', line, ival: 0 };
  }

  private parseUnary(): ASTNode {
    const line = this.cur.line;
    if (this.check(TokenType.MINUS) || this.check(TokenType.NOT)) {
      const op = this.advance();
      return { kind: 'UNOP', line, op: op.lexeme, left: this.parseUnary() };
    }
    return this.parsePrimary();
  }

  private parseMul(): ASTNode {
    let left = this.parseUnary();
    while (this.check(TokenType.STAR) || this.check(TokenType.SLASH) || this.check(TokenType.PERCENT)) {
      const op = this.advance();
      left = { kind: 'BINOP', line: left.line, op: op.lexeme, left, right: this.parseUnary() };
    }
    return left;
  }

  private parseAdd(): ASTNode {
    let left = this.parseMul();
    while (this.check(TokenType.PLUS) || this.check(TokenType.MINUS)) {
      const op = this.advance();
      left = { kind: 'BINOP', line: left.line, op: op.lexeme, left, right: this.parseMul() };
    }
    return left;
  }

  private parseRel(): ASTNode {
    let left = this.parseAdd();
    while (
      this.check(TokenType.LT) || this.check(TokenType.LE) ||
      this.check(TokenType.GT) || this.check(TokenType.GE)
    ) {
      const op = this.advance();
      left = { kind: 'BINOP', line: left.line, op: op.lexeme, left, right: this.parseAdd() };
    }
    return left;
  }

  private parseEq(): ASTNode {
    let left = this.parseRel();
    while (this.check(TokenType.EQ) || this.check(TokenType.NEQ)) {
      const op = this.advance();
      left = { kind: 'BINOP', line: left.line, op: op.lexeme, left, right: this.parseRel() };
    }
    return left;
  }

  private parseAnd(): ASTNode {
    let left = this.parseEq();
    while (this.check(TokenType.AND)) {
      const op = this.advance();
      left = { kind: 'BINOP', line: left.line, op: op.lexeme, left, right: this.parseEq() };
    }
    return left;
  }

  private parseOr(): ASTNode {
    let left = this.parseAnd();
    while (this.check(TokenType.OR)) {
      const op = this.advance();
      left = { kind: 'BINOP', line: left.line, op: op.lexeme, left, right: this.parseAnd() };
    }
    return left;
  }

  private parseAssign(): ASTNode {
    // Peek: if IDENT followed by '=', it's assignment.
    if (this.check(TokenType.IDENT)) {
      const savedLex = this.lex.saveState();
      const savedCur = this.cur;

      const ident = this.advance();
      if (this.check(TokenType.ASSIGN)) {
        this.advance(); // eat =
        return { kind: 'ASSIGN', line: ident.line, name: ident.lexeme, value: this.parseAssign() };
      }
      // Not assignment — restore and parse as or_expr.
      this.lex.restoreState(savedLex);
      this.cur = savedCur;
    }
    return this.parseOr();
  }

  private parseExpr(): ASTNode {
    return this.parseAssign();
  }

  // ----------------------------------------------------------------- statements

  private parseStmt(): ASTNode {
    const line = this.cur.line;

    // var declaration: int name;
    if (this.check(TokenType.KW_INT)) {
      this.advance();
      const name = this.expect(TokenType.IDENT);
      this.expect(TokenType.SEMICOLON);
      return { kind: 'VAR_DECL', line, name: name.lexeme };
    }

    // return
    if (this.match(TokenType.KW_RETURN)) {
      if (this.check(TokenType.SEMICOLON)) {
        this.advance();
        return { kind: 'RETURN', line, expr: null };
      }
      const expr = this.parseExpr();
      this.expect(TokenType.SEMICOLON);
      return { kind: 'RETURN', line, expr };
    }

    // if
    if (this.match(TokenType.KW_IF)) {
      this.expect(TokenType.LPAREN);
      const cond = this.parseExpr();
      this.expect(TokenType.RPAREN);
      const thenBranch = this.parseStmt();
      let elseBranch: ASTNode | null = null;
      if (this.match(TokenType.KW_ELSE)) {
        elseBranch = this.parseStmt();
      }
      return { kind: 'IF', line, cond, thenBranch, elseBranch };
    }

    // while
    if (this.match(TokenType.KW_WHILE)) {
      this.expect(TokenType.LPAREN);
      const cond = this.parseExpr();
      this.expect(TokenType.RPAREN);
      const body = this.parseStmt();
      return { kind: 'WHILE', line, cond, body };
    }

    // print(expr);
    if (this.match(TokenType.KW_PRINT)) {
      this.expect(TokenType.LPAREN);
      const val = this.parseExpr();
      this.expect(TokenType.RPAREN);
      this.expect(TokenType.SEMICOLON);
      return { kind: 'PRINT', line, expr: val };
    }

    // block
    if (this.check(TokenType.LBRACE)) {
      return this.parseBlock();
    }

    // expression statement
    const expr = this.parseExpr();
    this.expect(TokenType.SEMICOLON);
    return { kind: 'EXPR_STMT', line, expr };
  }

  private parseBlock(): ASTNode {
    const line = this.cur.line;
    this.expect(TokenType.LBRACE);
    const stmts: ASTNode[] = [];
    let guard = 0;
    while (!this.check(TokenType.RBRACE) && !this.check(TokenType.EOF)) {
      stmts.push(this.parseStmt());
      if (++guard > 100000) break; // safety valve
    }
    this.expect(TokenType.RBRACE);
    return { kind: 'BLOCK', line, stmts };
  }

  // ----------------------------------------------------------------- functions / program

  private isType(): boolean {
    return this.check(TokenType.KW_INT) || this.check(TokenType.KW_VOID);
  }

  private parseFunc(): ASTNode | null {
    const line = this.cur.line;
    if (!this.isType()) {
      this.error('expected type');
      return null;
    }
    this.advance();

    const name = this.expect(TokenType.IDENT);
    this.expect(TokenType.LPAREN);

    const params: string[] = [];
    if (!this.check(TokenType.RPAREN)) {
      do {
        if (!this.isType()) {
          this.error('expected parameter type');
          break;
        }
        this.advance(); // eat type
        const pname = this.expect(TokenType.IDENT);
        params.push(pname.lexeme);
      } while (this.match(TokenType.COMMA));
    }
    this.expect(TokenType.RPAREN);

    const body = this.parseBlock();

    return { kind: 'FUNC', line, name: name.lexeme, params, body };
  }

  parseProgram(): ASTNode {
    const funcs: ASTNode[] = [];
    while (!this.check(TokenType.EOF)) {
      const fn = this.parseFunc();
      if (!fn) break;
      funcs.push(fn);
    }
    return { kind: 'PROGRAM', line: 1, funcs };
  }
}
